core/common/src/main/java/alluxio/collections/TwoKeyConcurrentMap.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import static java.util.stream.Collectors.toSet;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * A two-level concurrent map implementation. Concurrency usage is controlled by the outer map.
+ *
+ * Users can supply the inner map type of their choice.
+ *
+ * @param <K1> the first key type
+ * @param <K2> the second key type
+ * @param <V> the value type
+ * @param <M> the type for the inner map
+ */
+public class TwoKeyConcurrentMap<K1, K2, V, M extends Map<K2, V>> extends ConcurrentHashMap<K1, M> {
+  private static final long serialVersionUID = 0L;
```

Comment:
NIT: is `0` correct?

Reply:
The number is arbitrary. It just gives us a way to notify java serialization when we make serialization-breaking changes to the class.